docs/design/plan-modification.md

```diff
@@ -0,0 +1,543 @@
+```go
+"base_image": {
+  Type:     schema.TypeString,
+  Required: true,
+  ForceNew: true,
+},
+```
+After receiving the `PlanResourceChange` request, the SDK determines whether any change was made to a field marked as `ForceNew`, and if so, adds it to the returned `RequiresNew` array.
+
+The SDK also executes logic at resource validation time (`InternalValidate`) to enforce the condition that if a resource does not have an Update function, all non-Computed attributes must have ForceNew set; and that if all fields are ForceNew or Computed without Optional, Update must _not_ be defined.
+
+### `DiffSuppressFunc`
+
+```go
+type SchemaDiffSuppressFunc func(k, old, new string, d *ResourceData) bool
+```
+
+Providers can use another schema behaviour, `DiffSuppressFunc`, to control whether a detected diff on a schema field should be considered semantically different. If this function returns true, any diff in the element values is ignored. This is commonly used to ignore differences in string capitalisation, or logically equivalent JSON values.
+
+### `CustomizeDiff`
+
+```go
+type CustomizeDiffFunc func(context.Context, *ResourceDiff, interface{}) error
+```
+
+Rather than exposing Terraform plans to provider developers, `helper/schema` has as a first-class concept the _resource diff_. Providers can optionally define a `CustomizeDiff` method on the `Resource` struct, which resembles a CRUD method, except that instead of `ResourceData` the function is supplied a `*ResourceDiff`, and is called during several points in the resource lifecycle, and must therefore be "resilient to support all scenarios".
+
+Unlike `DiffSuppressFunc`, `CustomizeDiff` is supplied the `meta` parameter, so API calls can be made.
+
+A large proportion of the examples of `CustomizeDiff` in large cloud provider code involves conditionally setting `ForceNew` behaviour on an attribute, most often:
+ - If certain conditions hold on the value of the attribute (e.g. if the bandwidth of an instance is reduced)
+ - If certain conditions hold on the values of the attribute and other attributes.
+
+`CustomizeDiff` is also used to implement multi-attribute validation, e.g. checking that at most one of two particular attributes is set. The present framework, on the other hand, separates plan modification from validation, and the solutions proposed below are concerned only with plan modification.
+
+### `helper/customdiff`
+
+The legacy SDK provides a set of reusable and composable helper functions in its `customdiff` package, intended to reduce the need for long and complex logic to reside in a single `CustomizeDiff` function in provider code. This small package is best summarised in its [documentation](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2/helper/customdiff).
+
+## Solution options
+
+### 1. `tfsdk.Resource.ModifyPlan()`
```

Comment:
Is another option here for providers to return a slice of plan modifiers that match an interface type? Then they can be composed together and sequencing is implicit. For example:

```go
type ResourcePlanModifier interface {
    Description(context.Context) string
    MarkdownDescription(context.Context) string
    ModifyPlan(context.Context, ModifyResourcePlanRequest, *ModifyResourcePlanResponse)
}

type ResourcePlanModifiers []ResourcePlanModifier

type ResourceWithPlanModifiers interface {
    Resource

    PlanModifiers(context.Context) ResourcePlanModifiers
}
```

This will allow the same documentation hooks provided by the `schema.Attribute.PlanModifiers` and let provider developers be a little more declarative:

```go
type customResource struct{}

type (r customResource) PlanModifiers(_ context.Context) ResourcePlanModifiers {
    return ResourcePlanModifiers{
        Modifier1,
        Modifier2,
    }
}
```